### DISCUSSION TITLE
New Installation from MPP-Solar on RaspianOS Bookworm on Raspbery Pi 3B+

### DISCUSSION THREAD

**Discussion by mmsfla** [jblance/mpp-solar#523] [General]
**Hints:**
Inverter: EASUN SPL5K (also labeld as POWLAND or EFFEKTA AX-M or K 5K)
Raspberry Pi: 3B+ (connect with USB-Cabel on EASUN)
RaspbianOS [since 12/2024]: Bookworm (based on Debian 12 Bookworm)
UseCase: I use the fabulous mppsolar-package to configuration of my equipment & monitor working conditions.
Security: Think about your Usecase! Expose this Constellation never ever in Internet! Use safe passwords for admin-users or use ssh-key-based procedures for operation system and web server applications, use also the transport layer security from grafana. Is your usecase bigger as the home wi-fi, you should it leave security-professionals. In particular, the database and Grafana should be examined more closely.

**Documentation:**
OS-Image-Installation  (Lite without Desktop and other) with Raspberry Pi Imager Tool on a 64 GB SDCARD (Don't forget the previous settings for Password, WLAN and SSH!)
After first login you should actualize your operating system.
Than you install by
`sudo apt install python3-full libsystemd-dev`
before you execute the pip-Installer. 
In my Envoiment i use the installation command
`sudo pip3 install --break-system-packages mppsolar[systemd]`

As next, i follow the original docu: https://github.com/jblance/mpp-solar/blob/3b500bd494477c1d3b0824823f8a43aa9a135711/docs/service-README.md

As next, i follow also the documentation for MQTT ...: https://github.com/jblance/mpp-solar/blob/3b500bd494477c1d3b0824823f8a43aa9a135711/docs/MQTT_Influx_Grafana.md

**Note 1 General:** 
both repos from grafana and influxdb have problems with gpg-Security, 
you should fix them, as the project docu describe:
https://packages.grafana.com/oss/
https://repos.influxdata.com/debian/

**Note 2 to InfluxDB:** before you use the database-Client _influx_ per command, you must install them with `sudo apt install influxdb-client`

sequal follows

**Comment by mmsfla** [comment]
Telegraf seems to have a problem with the INPUT parser.
Maybe someone will see it:

**Message sniffed on Mosquitto:**
```
EASUN/status/parallel_instance_number/value valid
EASUN/status/serial_number/value 55355535553555
EASUN/status/work_mode/value Line Mode
EASUN/status/fault_code/value No fault
EASUN/status/grid_voltage/value 231.7
EASUN/status/grid_voltage/unit V
EASUN/status/grid_frequency/value 49.97
... removed ...
EASUN/status/max_charger_range/unit A
EASUN/status/max_ac_charger_current/value 2
EASUN/status/max_ac_charger_current/unit A
EASUN/status/pv_input_current/value 0
EASUN/status/pv_input_current/unit A
EASUN/status/battery_discharge_current/value 0
EASUN/status/battery_discharge_current/unit A
```

I mean, all is normaly!

**Telegraf daemon status**

```
pi@raspberrypi3:~ $ sudo systemctl status telegraf.service 
● telegraf.service - Telegraf
     Loaded: loaded (/lib/systemd/system/telegraf.service; enabled; preset: enabled)
     Active: active (running) since Fri 2025-01-03 07:19:43 CET; 11h ago
       Docs: https://github.com/influxdata/telegraf
   Main PID: 42654 (telegraf)
      Tasks: 10 (limit: 767)
        CPU: 5min 7.816s
     CGroup: /system.slice/telegraf.service
             └─42654 /usr/bin/telegraf -config /etc/telegraf/telegraf.conf -config-directory /etc/telegraf/telegraf.d

Jan 03 19:03:44 raspberrypi3 telegraf[42654]: 2025-01-03T18:03:44Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "0"
Jan 03 19:03:44 raspberrypi3 telegraf[42654]: 2025-01-03T18:03:44Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "A"
Jan 03 19:03:44 raspberrypi3 telegraf[42654]: 2025-01-03T18:03:44Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "0"
Jan 03 19:03:44 raspberrypi3 telegraf[42654]: 2025-01-03T18:03:44Z **E! [inputs.mqtt_consumer]** Error in plugin: metric parse error: expected tag at 1:2: "A"
Jan 03 19:03:45 raspberrypi3 telegraf[42654]: 2025-01-03T18:03:45Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 03 19:03:55 raspberrypi3 telegraf[42654]: 2025-01-03T18:03:55Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 03 19:04:05 raspberrypi3 telegraf[42654]: 2025-01-03T18:04:05Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 03 19:04:15 raspberrypi3 telegraf[42654]: 2025-01-03T18:04:15Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 03 19:04:25 raspberrypi3 telegraf[42654]: 2025-01-03T18:04:25Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 03 19:04:35 raspberrypi3 telegraf[42654]: 2025-01-03T18:04:35Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
pi@raspberrypi3:~ $
```

This is complete!

**Telegraf configuration input/output**

```
pi@raspberrypi3:~ $ cat /etc/telegraf/telegraf.d/mqtt-input.conf && cat /etc/telegraf/telegraf.d/influx-output.conf 
[[inputs.mqtt_consumer]]
  servers = ["tcp://127.0.0.1:1883"]
  topics = [ 
    "#", 
  ]
  data_format = "influx"
  
[[outputs.influxdb]]
  urls = ["http://127.0.0.1:8086"]
  database = "mppsolar"
  skip_database_creation = true
  username = "xxx removed xxx"
  password = "xxx removed xxx"

pi@raspberrypi3:~ $
```

Using the troubleshooting commands I can get measurements into the database. It really seems to be a Telegraf input parser problem. The mpp-solar command is QPGS0 and the protocol is PI30.

I experimented with the input topic.
If I replaced `#` with `EASUN/settings/#` the INPUT plugin error did not appear.
However, the database had no content either.

I don't fully understand the documentary yet: https://github.com/influxdata/telegraf/tree/master/plugins/inputs/mqtt_consumer

But I think I need to look into it more.

But maybe someone had a similar problem?
After all, everything worked until the SD card broke.
I didn't have any other hardware at the time.

**Comment by jblance** [comment]
What output module are you using?
I think from memory telegraf needs a different message format

**Comment by jblance** [comment]
Ie try the influx2_mqtt output module

**Reply by mmsfla** [reply]
Thank you for your answer. I tried it.
But the telegraph service didn't start.
I think, the problem is my topic design.
I'll be dealing with it throughout the day today.

**Comment by jblance** [comment]
It shouldn't stop telegraf from starting - even if mqtt messages are
completely wrong.
What format messages did you get?
What errors did you get on telegraf?

**Reply by mmsfla** [reply]
wow, another answer!!! :-D

**Reply by mmsfla** [reply]
```
Jan 04 09:57:07 raspberrypi3 telegraf[67141]: 2025-01-04T08:57:07Z D! [agent] Successfully connected to outputs.influxdb
Jan 04 09:57:07 raspberrypi3 telegraf[67141]: 2025-01-04T08:57:07Z D! [agent] Starting service inputs
Jan 04 09:57:07 raspberrypi3 telegraf[67141]: 2025-01-04T08:57:07Z I! [inputs.mqtt_consumer] Connected [tcp://127.0.0.1:1883]
Jan 04 09:57:17 raspberrypi3 telegraf[67141]: 2025-01-04T08:57:17Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
~
~
~
pi@raspberrypi3:~ $ cat /etc/telegraf/telegraf.d/mqtt-input.conf && cat /etc/telegraf/telegraf.d/influx-output.conf 
[[inputs.mqtt_consumer]]
  servers = ["tcp://127.0.0.1:1883"]
  topics = [ 
    "/EASUN/status/#", 
  ]
  data_format = "influx"

[[outputs.influxdb]]
  urls = ["http://127.0.0.1:8086"]
  database = "mppsolar"
  skip_database_creation = true
  username = "grafana"
  password = "xxx"

pi@raspberrypi3:~ $ 
```

the telegram-daemon-output by start from this i/o configuration

any minutes later, if the first message pass the input

```

pi@raspberrypi3:~ $ sudo systemctl status telegraf.service 
● telegraf.service - Telegraf
     Loaded: loaded (/lib/systemd/system/telegraf.service; enabled; preset: enabled)
     Active: active (running) since Sat 2025-01-04 09:57:07 CET; 8min ago
       Docs: https://github.com/influxdata/telegraf
   Main PID: 67141 (telegraf)
      Tasks: 9 (limit: 767)
        CPU: 3.063s
     CGroup: /system.slice/telegraf.service
             └─67141 /usr/bin/telegraf -config /etc/telegraf/telegraf.conf -config-directory /etc/telegraf/telegraf.d

Jan 04 10:03:37 raspberrypi3 telegraf[67141]: 2025-01-04T09:03:37Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 04 10:03:47 raspberrypi3 telegraf[67141]: 2025-01-04T09:03:47Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 04 10:03:57 raspberrypi3 telegraf[67141]: 2025-01-04T09:03:57Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 04 10:04:07 raspberrypi3 telegraf[67141]: 2025-01-04T09:04:07Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 04 10:04:17 raspberrypi3 telegraf[67141]: 2025-01-04T09:04:17Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 04 10:04:27 raspberrypi3 telegraf[67141]: 2025-01-04T09:04:27Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 04 10:04:37 raspberrypi3 telegraf[67141]: 2025-01-04T09:04:37Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 04 10:04:47 raspberrypi3 telegraf[67141]: 2025-01-04T09:04:47Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 04 10:04:57 raspberrypi3 telegraf[67141]: 2025-01-04T09:04:57Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
Jan 04 10:05:07 raspberrypi3 telegraf[67141]: 2025-01-04T09:05:07Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
pi@raspberrypi3:~ $ 

```

the known signal from output, but what is in database ?

```
pi@raspberrypi3:~ $ influx
Connected to http://localhost:8086 version 1.6.7~rc0
InfluxDB shell version: 1.6.7~rc0
> use mppsolar
Using database mppsolar
> SHOW MEASUREMENTS
> 
```
... nothing ...

**Reply by mmsfla** [reply]
Which function have the "#" - sign  in topic-directive?
Is this a placeholder for the payload from the inverter?

**Comment by jblance** [comment]
It's a wildcard for mqtt topics, in this instance it means all topics that
start with /EASUN/status no matter how many more parts there are

**Comment by jblance** [comment]
What did you change that stopped the telegraf service?
I was talking about changes at the mppsolar end of the pipeline

**Reply by mmsfla** [reply]
First I used Telegraf as documented.
The Input-MQTT plugin only had the wildcard **"#"** as a topic.
This causes the plugin to crash and outputs line after line of the message as an error of the plugin. 
See here:
```
pi@raspberrypi3:~ $ tail -f /var/log/telegraf/telegraf.log 
2025-01-04T13:56:05Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "A"
2025-01-04T13:56:05Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:4: "110"
2025-01-04T13:56:05Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "A"
2025-01-04T13:56:05Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "2"
2025-01-04T13:56:05Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "A"
2025-01-04T13:56:05Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "0"
2025-01-04T13:56:05Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "A"
2025-01-04T13:56:05Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "0"
2025-01-04T13:56:05Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "A"
2025-01-04T13:56:08Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:6: "valid"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:15: "55355535553555"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected field at 1:10: "Line Mode"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected field at 1:9: "No fault"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:6: "233.4"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "V"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:6: "50.01"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:3: "Hz"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:6: "233.4"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:2: "V"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:6: "50.01"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:3: "Hz"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error in plugin: metric parse error: expected tag at 1:4: "700"
2025-01-04T13:56:11Z E! [inputs.mqtt_consumer] Error i ....
```
If i change the topic to "/EASUN/status/#", this error disappears. 
However, nothing arrives in the database.
I'm sniffing with tcpdump on localhost tcp/8086.
Nothing arrives this port.
So the outport-plugin or the data-processor doesn't work. I don't know.

Either this is a bug in the current version (actually version 1.33)!
In Telegraf versions 1.26 and 1.28 the configuration with topic "#" worked without any problems.
Since these versions, some changes have been made to the MQTT consumer plugin. Maybe the parser now expects something different than what I give it with "mpp-solar -c QPGS0". I don't know.

Here the changelog from telegraf: https://raw.githubusercontent.com/influxdata/telegraf/master/CHANGELOG.md
That's my modest research. I'm not competent enough when it comes to coding the plugin.

My issue by influxdata: https://github.com/influxdata/telegraf/issues/16370
